WIP: Docker containers

mcarpenterjr opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here unless your commit contains the issue number -->

Add support for Docker containers, this PR does/will contain a Dockerfile and matching docker-compose.yml for use in the creation and deployment of a Suite CRM container. 

It desires to be ambiguous so it can be used from the CLI to create an image or deploy a container as well as support the needs of automated deployment through a CI if needed.

At the lowest level simply running `docker build ./docker/` from the root of Suite should yield a fresh image built from the "master" branch of the official repository. 

At the highest (That I've thought of so far) using docker-compose, a user should be able to pass a target repository, branch and specify if the container should use an internal database or a separate container for the database.

No matter the level of abstraction Suite should install and set itself up, so when a container is started Suite is ready to rock.


## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

This breaks the dependence on third parties for the supply of containers, also will help with quick deployment of an instance.

In the running instance I am working from, gitlab i used to track any changes, but has to rsync anything done through the CI to the container that is host to suite. The container is an image from the third party bitnami (I have nothing against Bitnami, just the fact that I'm at their mercy when it comes to Suite) I've focused on the gitlab-ci to get started but this should work with any scriptable continuous integration out there. Or right from the command line.

## How To Test This
<!--- Please describe in detail how to test your changes. -->

You should have both docker and docker-compose installed.

Running `docker-compose -f ./docker/docker-compose.yml up -d --force-recreate --always-recreate-deps` from the branches root will create a mariadb container and then build and deploy the docker container.

Running `docker build --force-rm -t smt/suitecrm:latest ./docker/` Will build an image that can then be used to deploy a container.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Final checklist
<!--- Go over all the following points and check all the boxes that apply. --->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! --->
- [ ] My code follows the code style of this project found [here](https://docs.suitecrm.com/community/contributing-code/coding-standards/).
- [X] My change requires a change to the documentation.
- [X] I have read the [**How to Contribute**](https://docs.suitecrm.com/community/contributing-code/) guidelines.

<!--- Your pull request will be tested via Travis CI to automatically indicate that your changes do not prevent compilation. --->

<!--- If it reports back that there are problems, you can log into the Travis system and check the log report for your pull request to see what the problem was. --->

In summary, at this point, the concept works, if docker is available the containers get created but the silent install part fails, and the DB conficuration is incomplete.

isleshocky77 committed:
Adding the abily to relate a Note with a Campaign

isleshocky77 committed:
PSR-2 Convert tabs to spaces

isleshocky77 committed:
Code Format after tab -> spaces conversion

isleshocky77 committed:
Remove duplicate array key (clean warning)

Dillon-Brown committed:
Remove fullcalendar

Dillon-Brown committed:
Fixed #3057 - FullCalendar update to version 3.1

Dillon-Brown committed:
Fixed #4511 - Upgrade of TinyMCE to latest 4.7.x version

Dillon-Brown committed:
Remove tinyMCE

Dillon-Brown committed:
Fixed #3100 - Upgrade for HTML Purifier [Request]

Dillon-Brown committed:
Merge pull request #6714 from gymad/develop_elasticpager_fix

fix last page when modulus is zero

Dillon-Brown committed:
Merge branch 'develop' into tiny_mce

Dillon-Brown committed:
Merge pull request #6730 from Dillon-Brown/issue-6720

Fixed #6720 - Remove old recaptcha files

Dillon-Brown committed:
Merge branch 'develop' into fullcalendar

Dillon-Brown committed:
Merge branch 'develop' into fullcalendar

isleshocky77 committed:
Cleaning up the clean cache command

* Update command to allow for cache directory specified in config
* Remove warnings/errors when no cached directory present
* Update to have namespace "cache" so we can add other cache commands later
  * Leave alias for current command
* Use provided toolsets required by robo for implementation
* Print cleaned directories
* Confirm before cleaning directory
  * Allow for force override

Dillon-Brown committed:
Merge branch 'develop' into fix-clean-cache-command-config-cache-dir

isleshocky77 committed:
Code formatting

Dillon-Brown committed:
Merge remote-tracking branch 'upstream/develop' into feature/campaign-notes

# Conflicts:
#	modules/Campaigns/Campaign.php
#	modules/Campaigns/views/view.detail.php
#	modules/Notes/vardefs.php

Dillon-Brown committed:
Reformat

Dillon-Brown committed:
Fix test

samus-aran committed:
Merge pull request #4606 from isleshocky77/feature/campaign-notes

Add ability to have Notes on a Campaign

Dillon-Brown committed:
Formatting

Dillon-Brown committed:
Merge pull request #6802 from isleshocky77/fix-clean-cache-command-config-cache-dir

Allow for clean cache command to use config'd cache_dir (plus clean up)

lazka committed:
Fixes #7111 - composer: include mpdf/mpdf 6.1 and drop in-tree version (5.7)

Remove the included version of mpdf (in modules/AOS_PDF_Templates/PDF_Lib) and
add it to composer instead.

This also updates the version from 5.7.1 to 6.1.0 to get some fixes that were already
applied manually to the version in SuiteCRM. There are many newer versions available,
but they require more changes and this is a good first step (and this potentially
could be backported to 7.10)

The version in tree had one manual fix applied (d1279cad9c717) which is included in
6.1.0 already, so the upstream version can be used as is.

The required changes to make things compatible with with 6.1 were
* mPDF_exception is now called MpdfException
* SetAutoFont() no longer exists, the docs suggest to use autoLangToFont instead
  to get the same results as before.

Fixes #7111

Dillon-Brown committed:
Merge remote-tracking branch 'upstream/develop' into fullcalendar

# Conflicts:
#	composer.json
#	composer.lock

Dillon-Brown committed:
Merge remote-tracking branch 'upstream/hotfix' into develop_merge

code-ph0y committed:
Merge pull request #6702 from Dillon-Brown/fullcalendar

Fixed #3057 - FullCalendar update to version 3.1

Dillon-Brown committed:
Merge branch 'develop' into develop_merge

samus-aran committed:
Merge pull request #6703 from Dillon-Brown/htmlpurifier

Fixed #3100 - Upgrade for HTML Purifier [Request]

Dillon-Brown committed:
Merge remote-tracking branch 'upstream/master' into develop_merge

Dillon-Brown committed:
Merge remote-tracking branch 'origin/develop_merge' into develop_merge

# Conflicts:
#	composer.json

Dillon-Brown committed:
Update composer.lock

Dillon-Brown committed:
Merge pull request #7131 from Dillon-Brown/develop_merge

Develop merge

Dillon-Brown committed:
Add pclzip/pclzip to composer

Dillon-Brown committed:
Remove pclzip references

Dillon-Brown committed:
Remove pclZip from include

Dillon-Brown committed:
Update composer.lock

Dillon-Brown committed:
Remove zip_utils.php

Dillon-Brown committed:
Reformat php_zip_utils.php

Dillon-Brown committed:
Update composer

Dillon-Brown committed:
Merge remote-tracking branch 'upstream/develop' into tiny_mce

# Conflicts:
#	composer.json
#	composer.lock

Dillon-Brown committed:
Update composer

Dillon-Brown committed:
Update tinymce references

Dillon-Brown committed:
Add tedivm/jshrink to composer

Dillon-Brown committed:
Remove Minifier.php

Dillon-Brown committed:
Fixed #7027 - Use a maintained library to minify JS

Dillon-Brown committed:
Merge remote-tracking branch 'upstream/develop' into fix-mpdf-composer

# Conflicts:
#	composer.json
#	composer.lock

Dillon-Brown committed:
Update composer

Dillon-Brown committed:
PHP 5.6 compat

code-ph0y committed:
Merge pull request #6699 from Dillon-Brown/tiny_mce

Fixed #4511 - Upgrade of TinyMCE to latest 4.9.x version

Dillon-Brown committed:
Merge branch 'develop' into fix-mpdf-composer

Dillon-Brown committed:
Merge pull request #7118 from lazka/fix-mpdf-composer

Fixes #7111 - composer: include mpdf/mpdf 6.1 and drop in-tree version (5.7)

code-ph0y committed:
Merge branch 'develop' into issue-7027

cameronblaikie committed:
Merge pull request #7295 from Dillon-Brown/issue-7027

Fixed #7027 - Use a maintained library to minify JS

Commit:
fixes #7308 Sub-Theme changes don't always update

Refactored most of the Robo build command for rebuillding sub themes

Commit:
FEATURE: Custom SMT THEME

Based on SuiteP

Commit:
FEATURE: Gitlab CI

Commit:
GITLABCI: Ty SSH

Commit:
GITLABCI: Update SSH Keys

Commit:
GITLABCI: Test Dir and User Sturcture

Commit:
GITLABCI: Troubleshooting

Dillon-Brown committed:
Merge remote-tracking branch 'upstream/develop' into feature/plcZip

# Conflicts:
#	composer.json
#	composer.lock

Dillon-Brown committed:
Update composer.lock

Dillon-Brown committed:
Merge remote-tracking branch 'upstream/develop' into feature/plcZip

# Conflicts:
#	composer.lock

Dillon-Brown committed:
Update composer.lock

Commit:
FEATURE: ContainerIzation

Coupled with creating a container we also need a method to silently install suite
on creation.

Commit:
FEATURE: Composer In Docker File.

Commit:
REFACTOR: MOAR DOCKER!

Updating build libraries for image.

Dillon-Brown committed:
Merge pull request #7287 from Dillon-Brown/feature/plcZip

Remove PclZip

Commit:
REFACTOR: more docker deps...

Added dev modules for zip, curl, png, jpeg and 1g...???

Commented out the configure line for the php extension gd.

broke up and correctd the collection and install of suite.

Commit:
REFACTOR: Configure composer install.

Get composer installing as expected, needed to specify `--working-dir=''` option.

cd into suite directory before executing the install script.

Commit:
FEATURE: Automated docker BUILDS...???

Commit:
REFACTOR: Make CI valid yml

Commit:
REFACTOR: Tell docker where the Dockerfile is.

Commit:
REFACTOR: make default password, password.

Commit:
REFACTOR: Add Support Files

Commit:
REFACTOR: Remove Whitespace

Commit:
REFACTOR: Server keys

Added HTTPS key for better support

Commit:
REFACTOR: GitlabCI Docker Builds

Commit:
REFACTOR: CI Build ENVS

Added some environment variables to help build the right version.

Commit:
BUGFIX: Spelling counts

Commit:
CI-BUGFIX: Use apache2 restart

over using reload.

Commit:
CI-BUGFIX: Cache Dirs

Commit:
CI-BUGFIX: Create Stages

Commit:
CI-BUGFIX: Cache Dirs

Commit:
CI-REFACTOR: ADD Perms on Suite Dir.

Commit:
REFACTOR: Remove config overrides from chmod.

Commit:
REFACTOR: Add Ambiguity to Docker

mcarpenterjr committed:
CI-DOCKER-BUGFIX: Update Repo Aegs

mcarpenterjr committed:
BUGFIX: Bad Vhost DIRS

Commit:
REFACTOR: Clean up the .gitignore

Added some todo's

Commit:
FEATURE: Docker Compose?

Commit:
CI-BGUFIX: Docker-Compose work around

Commit:
CI-BUGFIX: APK not a thing in gitlab-runner

Commit:
CI-BUGFIX: Python user

Commit:
CI-BUGFIX: Indentaion Mattters

Commit:
CI-BUGFIX: Pip install with user

Commit:
CI-BUGFIX: Find the compose file

Commit:
CI-BUGFIX: specify docker compose file.

Commit:
CI-BUGFIX: Volumes Aren't part of services

Commit:
CI-BUGFIX: Mariadb is just mariadb

Commit:
CI-BUGFIX: Exposing ports

Commit:
CI_BUGFIX: Detach compose after start.

Commit:
CI-BUGFIX: Update DB ENVs

Commit:
CI-BUGFIX: --detach not a docker compose thing

Commit:
REFACTOR: ENVs in docker build pipeline.

Commit:
CI-REFACTOR: -e not a thing

Moved ENV creation for compose deeper into the ci and not
part of the docker-compose command

Commit:
CI-REFACTOR: ENVS

Commit:
CI-REFACTOR: environment should be mapping not array

connorshea commented:
It looks like this has a bunch of extra changes that aren't relevant, e.g. the CSS changes. Besides the addition of the Dockerfile, is any of this strictly necessary?

Commit:
CI-REFACTOR: No ENVs

mcarpenterjr commented:
Nope, it should be pared down some, probably should have synced my gitlab repo with GitHub before creating the Draft.

Commit:
Merge branch 'develop' of https://github.com/salesagility/SuiteCRM into develop

Commit:
MERGE: Develop

Bringing all dep up to speed

Commit:
REFACTOR: Docker Envs, SMT Specifics

Remove the SMT CSS

Add config_si.php and update the related ENVs.

Commit:
BUGFIX: Missing imap Dep.

Commit:
REFACTOR: show build source in CI

Commit:
REFACTOR: Show Branch in CI

Commit:
BUGFIX: Get Branch

Spelling counts.

Commit:
BUGFIX: Docker IMAP

Commit:
BUGFIX: Deps install, bad syntax

Commit:
BUGFIX: imap deps.

Commit:
REFACTOR: ignor custom themes...

Commit:
REFACTOR: PHP7?

Commit:
REFACTOR: Change database type

Commit:
BUGFIX: ENABLE mysql in php

Commit:
CI-BUGFIX: PHP Mysqli

Commit:
CI-BUGFIX: Mysqli

Commit:
CI-REFACTOR: mysqli...

Commit:
CI-BUGFIX: php-mysql...?

Commit:
CI-REFACTOR: PHP extensions

UGH! PHP extenision in docker don't have installation candidates, they have
special install methods.

Commit:
CI-REFACTOR: Passwords are madatory

The mariadb container's `MYSQL_ROOT_PASSWORD` variable is required.
Larter this should be updated to either let the container generate a
random one, or have the option for the build pipe pass one.

We're also adding support for a mysql user password, currently
one can either be passed or not, allow empty passwords is on.

Commit:
CI-REFACTOR: Rely on Setup for MySQL Setup

Create DB and Demo Data durin SI.

Commit:
CI-REFACTOR: Configure suite.

Suite seems to be launching and running on completion, but doesn't seem to be getting configured.

Commit:
CI-AUTOINSTALL: Just run the install script with no ARGS.

Commit:
CI-INSTALL: No aut install.

Commit:
REFACTOR: Remove Custom Stuff

Re-Enabled Silent install

Commit:
CI-REFACTOR: Use sample Config

Commit:
CI-REFACTOR: No Auto Install

Commit:
CI-REFACTOR: Add Bakc Auto Install

Commit:
CI-REFACTOR: Volumes?

Commit:
CI-REFACTOR: Use alternate config file.

Commit:
CI-REFACTOR: sub in known DB

Commit:
CI-REFACTOR: match config to working travis_ci

Commit:
REFACOTR: $_ENV does not work in scripts

Commit:
REFACTOR: No condition on DB settings

Commit:
REFACTOR: Sub Out `$_ENV` for `getenv()`

Commit:
CI-REFACTOR: ENV's not expanding, trying
hard coded passwords

Commit:
DOCKER-REFACTOR: Expose MYSQL port

Commit:
DOCKER-REFACTPR: Reduce Image Layers

Purge old data

Commit:
CI-REFACTOR: Fix Compose

Commit:
CI-REFACTOR: Passwords?

Commit:
REFACTOR: install.php

Commit:
REFACTOR: CI

Commit:
REFACTOR: CI

DOCKER-REFACTOR: Debugging, add Xdebug into the mix.

DOCKER-REFACTOR: Adjust xdebug Settings.

DOCKER-REFACTOR: Bool has to be string?

DOCKER-REFACTOR: Expose xdebug port

DOCKER-REFACTOR: Expose 9000 on Dockerfile also

DOCKER-REFACTOR: Map xdebug port.

Commit:
DOCKER-REFACTOR: Debugging, add Xdebug into the mix.

Commit:
DOCKER-REFACTOR: Adjust xdebug Settings.

Commit:
DOCKER-REFACTOR: Bool has to be string?

Commit:
DOCKER-REFACTOR: Expose xdebug port

Commit:
DOCKER-REFACTOR: Expose 9000 on Dockerfile also

Commit:
DOCKER-REFACTOR: Map xdebug port.

Commit:
DOCKER-REFACTOR: Update xdebug config

Had to exclusively specify my IP, this needs to get refactored out to
an ENV variable for those looking to remote debug.

Commit:
DOCKER-REFACTOR: Update xdebug config

Had to exclusively specify my IP, this needs to get refactored out to
an ENV variable for those looking to remote debug.

Commit:
REFACTOR: Make `clean_string()` a touch more verbose

Commit:
REFACTOR: Make `clean_string()` a touch more verbose

REFACTOR: Stop watching my vscode folder

REFACTOR: Organise file

Commit:
REFACTOR: Stop watching my vscode folder

Commit:
REFACTOR: Organise file

Commit:
REFACTOR: Stop watching dev files,
Directory housekeeping.

Commit:
REFACTOR: Stop watching dev files

Commit:
DOCKER-REFACTOR: Reduce Image Size

Attempting to reduce number of layers currently
we are at 1.5G up from 375M of the base.

REFACTOR: SSL Certs

DOCKER-REFACTOR: Make openssl love passwords again.

DOCKER-BUGFIX: Error in SSL Run Sequence

DOCKER-BUGFIX: SSL Country Name only 2 chars

DOCKER-REFACTOR: Update Apache Settings.

Commit:
REFACTOR: Reduce Images

Commit:
REFACTOR: SSL Certs

Commit:
DOCKER-REFACTOR: Make openssl love passwords again.

Commit:
DOCKER-BUGFIX: Error in SSL Run Sequence

Commit:
DOCKER-BUGFIX: SSL Country Name only 2 chars

Commit:
DOCKER-REFACTOR: Update Apache Settings.

Commit:
DOCKER-BUGFIX: Command Chaining

Commit:
DOCKER-BUGFIX: Hold off on HTTPS

Commit:
DOCKER-BUGFIX: Hold off on HTTPS

DOCKER-BUGFIX: Command Chaining

DOCKER-REFACTOR: Reduce Duplicate Dependancies

DOCKER-REFACTOR: libxml2-dev Required

DOCKER-REFACTOR: zlib1g-dev Required

Commit:
DOCKER-REFACTOR: Reduce Duplicate Dependancies

Commit:
DOCKER-REFACTOR: libxml2-dev Required

Commit:
DOCKER-REFACTOR: zlib1g-dev Required

Commit:
DOCKER-REFACTOR: Image Size

The image is as small as possible at this point in time.
Split the build proccess back into multiple layers.

SSL Still needs configuration.

Commit:
DOCKER-REFACTOR: Image Size

The image is as small as possible at this point in time.
Split the build proccess back into multiple layers.

SSL Still needs configuration.

Commit:
Merge branch 'docker-containers' of dev2:dev/suite-crm into docker-containers

Commit:
DOCKER-FEATURE: Turn HTTPS back on.

Commit:
DOCKER-BUGFIX: Typo

Commit:
DOCKERFILE-REFACTOR: move generated key and cert

Commit:
DOCKER-REFCTOR: cert locations

Try and create them in the suitecrm root directory.

Commit:
COMMIT FORGOTTEN CHANGES

Commit:
DOCKER-REFACTOR: make cert dir

Commit:
DOCKER-REFACTOR: More cert locations and copy action

Commit:
DOCKER-REFATOR: Order Of Operations

Commit:
DOCKER-REFACTOR: Certs

Generate certs and just tell the HTTPS vhost to look in the suite/docker dir

Commit:
DOCKER-REFACTOR: Give the HTTPS vhost the right files.

Commit:
DOCKER-REFACTOR: DB Envs

Update Database Credentials.

Commit:
DOCKER-REFACTOR: Update Setup ENVS

Pull ENVS Even Further Through

Commit:
DOCKER-REFACTOR: Clarify DM Root password From DB User password

gardner commented:
This would be great for local development and production deployment. Is there anything that I can do to help?

isleshocky77 commented:
It looks like you have some changes as part of this PR that should be broken into separate PRs to make it more manageable to test and merge in.

pgorod commented:
@gil0109 note that this PR is marked as `Draft`, unfinished. So we're not considering this for review yet.

I hope @mcarpenterjr can finish it soon! Thanks

2xaronl commented:
I've just started looking into containerizing SuiteCRM and I found this pull request.  It looks like a LOT of effort was put into this years ago.  Any chance of @mcarpenterjr picking this back up or someone else jumping in?
